jina/peapods/runtimes/gateway/graph/topology_graph.py

```diff
@@ -0,0 +1,135 @@
+import asyncio
+
+from collections import defaultdict
+from typing import List, Optional, Dict, Tuple
+
+
+class TopologyGraph:
+    """
+    :class TopologyGraph is a class that describes a computational graph of nodes, where each node represents
+        a Pod that needs to be sent messages in the order respecting the path traversal.
+
+    :param graph_description: A dictionary describing the topology of the Pods. 2 special nodes are expected, the name `start-gateway` and `end-gateway` to
+        determine the nodes that receive the very first message and the ones whose response needs to be sent back to the client. All the nodes with no outgoing nodes
+        will be considered to be hanging, and they will be "flagged" so that the user can ignore their tasks and not await them.
+    """
+
+    class _ReqReplyNode:
+        def __init__(self, name: str, number_of_parts: int = 1, hanging: bool = False):
+            self.name = name
+            self.outgoing_nodes = []
+            self.number_of_parts = number_of_parts
+            self.hanging = hanging
+            self.parts_to_send = []
+
+        @property
+        def leaf(self):
+            return len(self.outgoing_nodes) == 0
+
+        async def _wait_previous_and_send(
+            self, msg, previous_task: Optional[asyncio.Task], connection_pool
+        ):
+            if previous_task is not None:
+                msg = await previous_task
+            if msg is not None:
+                self.parts_to_send.append(msg)
+                # this is a specific needs
+                if len(self.parts_to_send) == self.number_of_parts:
+                    resp = await connection_pool.send(self.parts_to_send[-1], self.name)
+                    return resp
+
+        def get_leaf_tasks(
```

Comment:
in the graph context the name of the function is of course correct, but may be a bit much detail for a caller? A caller might just do 'send the message to this node in the graph and get me back the task'? Not sure the caller is interested in the fact about leaf tasks etc? But no strong opinion on this

Reply:
Yes, it may be true. I just thought since implementation is complex to show what it really does. 

U can propose names